File: eel-core/src/main/scala/io/eels/component/parquet/ParquetSource.scala
package io.eels.component.parquet

import com.sksamuel.exts.Logging
import com.sksamuel.exts.OptionImplicits._
import com.sksamuel.exts.io.Using
import io.eels.datastream.Publisher
import io.eels.{Predicate, _}
import io.eels.schema.StructType
import org.apache.hadoop.conf.Configuration
import org.apache.hadoop.fs.{FileSystem, Path}
import org.apache.parquet.format.converter.ParquetMetadataConverter
import org.apache.parquet.hadoop.{Footer, ParquetFileReader}

import scala.collection.JavaConverters._

object ParquetSource {

  def apply(string: String)(implicit fs: FileSystem, conf: Configuration): ParquetSource = apply(FilePattern(string))

  def apply(uri: java.net.URI)(implicit fs: FileSystem, conf: Configuration): ParquetSource = apply(FilePattern(new Path(uri.toString)))

  def apply(path: java.nio.file.Path)(implicit fs: FileSystem, conf: Configuration): ParquetSource = apply(FilePattern(path))

  def apply(path: Path)(implicit fs: FileSystem, conf: Configuration): ParquetSource = apply(FilePattern(path))
}

case class ParquetSource(pattern: FilePattern,
                         predicate: Option[Predicate] = None,
                         projection: Seq[String] = Nil,
                         dictionaryFiltering: Boolean = true,
                         caseSensitive: Boolean = true)
                        (implicit fs: FileSystem, conf: Configuration) extends Source with Logging with Using {
  logger.debug(s"Created parquet source with pattern=$pattern")

  lazy val paths: List[Path] = pattern.toPaths()

  def withDictionaryFiltering(dictionary: Boolean): ParquetSource = copy(dictionaryFiltering = dictionary)
  def withCaseSensitivity(caseSensitive: Boolean): ParquetSource = copy(caseSensitive = caseSensitive)
  def withPredicate(pred: => Predicate): ParquetSource = copy(predicate = pred.some)
  def withProjection(first: String, rest: String*): ParquetSource = withProjection(first +: rest)
  def withProjection(fields: Seq[String]): ParquetSource = {
    require(fields.nonEmpty)
    copy(projection = fields.toList)
  }

  // returns the metadata in the parquet file, or an empty map if none
  def metadata(): Map[String, String] = {
    paths.foldLeft(Map.empty[String, String]) { (metadata, path) =>
      val footer = ParquetFileReader.readFooter(conf, path, ParquetMetadataConverter.NO_FILTER)
      metadata ++ footer.getFileMetaData.getKeyValueMetaData.asScala
    }
  }

  // todo should take the merged schema from all files
  lazy val schema: StructType = RowParquetReaderFn.schema(paths.headOption.getOrError("No paths found for source"))

  // returns the count of all records in this source, predicate is ignored
  def countNoPredicate(): Long = statistics().count

  // returns stats, predicate is ignored
  def statistics(): Statistics = {
    if (paths.isEmpty) Statistics.Empty
    else {
      paths.foldLeft(Statistics.Empty) { (stats, path) =>
        val footer = ParquetFileReader.readFooter(conf, path, ParquetMetadataConverter.NO_FILTER)
        footer.getBlocks.asScala.foldLeft(stats) { (stats, block) =>
          stats.copy(
            count = stats.count + block.getRowCount,
            compressedSize = stats.compressedSize + block.getCompressedSize,
            uncompressedSize = stats.uncompressedSize + block.getTotalByteSize
          )
        }
      }
    }
  }

  override def parts(): Seq[Publisher[Seq[Row]]] = {
    logger.debug(s"Parquet source has ${paths.size} files: ${paths.mkString(", ")}")
    paths.filter(it => !it.getName.equalsIgnoreCase(s"_SUCCESS") && !it.getName.startsWith(".")).map
    { it => new ParquetPublisher(it, predicate, projection, caseSensitive, dictionaryFiltering) }
  }

  def footers(): List[Footer] = {
    logger.debug(s"Parquet source will read footers from $paths")
    paths.flatMap { it =>
      val status = fs.getFileStatus(it)
      logger.debug(s"status=$status; path=$it")
      ParquetFileReader.readAllFootersInParallel(fs.getConf, status).asScala
    }
  }
}

File: eel-core/src/main/kotlin/io/eels/component/parquet/ParquetSource.kt
package io.eels.component.parquet

import io.eels.FilePattern
import io.eels.Row
import io.eels.Source
import io.eels.component.Part
import io.eels.component.Using
import io.eels.component.avro.AvroSchemaFns
import io.eels.component.avro.AvroSchemaMerge
import io.eels.schema.Schema
import io.eels.util.Logging
import io.eels.util.Option
import io.eels.util.getOrElse
import org.apache.hadoop.conf.Configuration
import org.apache.hadoop.fs.FileSystem
import org.apache.hadoop.fs.Path
import org.apache.parquet.hadoop.Footer
import org.apache.parquet.hadoop.ParquetFileReader
import rx.Observable

data class ParquetSource @JvmOverloads constructor(val pattern: FilePattern,
                                                   val fs: FileSystem = FileSystem.get(Configuration()),
                                                   val configuration: Configuration = Configuration()) : Source, Logging, Using {

  @JvmOverloads constructor(path: java.nio.file.Path,
                            fs: FileSystem = FileSystem.get(Configuration()),
                            configuration: Configuration = Configuration()) : this(FilePattern(path, fs), fs, configuration)

  @JvmOverloads constructor(path: Path,
                            fs: FileSystem = FileSystem.get(Configuration()),
                            configuration: Configuration = Configuration()) : this(FilePattern(path, fs), fs, configuration)

  // the schema returned by the parquet source should be a merged version of the
  // schemas contained in all the files.
  override fun schema(): Schema {
    val paths = pattern.toPaths()
    val schemas = paths.map { path ->
      using(ParquetReaderFns.createReader(path, Option.None, Option.None)) { reader ->
        Option(reader.read()).getOrElse { error("Cannot read $path for schema; file contains no records") }.schema
      }
    }
    val avroSchema = AvroSchemaMerge("record", "namspace", schemas)
    return AvroSchemaFns.fromAvroSchema(avroSchema)
  }

  override fun parts(): List<Part> {
    val paths = pattern.toPaths()
    logger.debug("Parquet source will read from $paths")
    val schema = schema()
    return paths.map { ParquetPart(it, schema) }
  }

  fun footers(): List<Footer> {
    val paths = pattern.toPaths()
    logger.debug("Parquet source will read from $paths")
    return paths.flatMap {
      val status = fs.getFileStatus(it)
      logger.debug("status=$status; path=$it")
      ParquetFileReader.readAllFootersInParallel(configuration, status)
    }
  }

  class ParquetPart(val path: Path, val schema: Schema) : Part {
    override fun data(): Observable<Row> = Observable.create { sub ->
      try {
        sub.onStart()
        val reader = ParquetReaderFns.createReader(path, Option.None, Option.None)
        ParquetRowIterator(reader).forEach {
          sub.onNext(it)
        }
      } catch (t: Throwable) {
        sub.onError(t)
      }
      if (!sub.isUnsubscribed)
        sub.onCompleted()
    }
  }
}
